File: wear/src/main/java/org/dwallach/xstopwatch/SharedState.java
/*
 * XStopwatch / XTimer
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/xstopwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/xstopwatch/licensing.html
 */
package org.dwallach.xstopwatch;

import android.app.Activity;
import android.content.Context;
import android.text.format.DateUtils;
import android.util.Log;

import java.util.Observable;

abstract class SharedState extends Observable {
    private final static String TAG = "SharedState";

    protected boolean running;
    protected boolean reset;
    protected long updateTimestamp;  // when the last user interaction was
    protected boolean visible;
    protected boolean initialized;

    protected SharedState() {
        running = false;
        reset = true;
        visible = false;
        initialized = false;
    }

    public long getUpdateTimestamp() {
        return updateTimestamp;
    }

    private void makeUpdateTimestamp() {
        updateTimestamp = currentTime();
    }

    public boolean isInitialized() {
        return initialized;
    }

    public void setVisible(boolean visible) {
        Log.v(TAG, getShortName() + "visible: " + visible);
        this.visible = visible;
        initialized = true;

        makeUpdateTimestamp();
        pingObservers();
    }

    public boolean isVisible() {
        return visible;
    }

    public static long currentTime() {
        return System.currentTimeMillis();
    }

    public boolean isRunning() {
        return running;
    }

    public boolean isReset() {
        return reset;
    }

    public void reset(Context context) {
        Log.v(TAG, getShortName() + "reset");
        running = false;
        reset = true;
        initialized = true;

        makeUpdateTimestamp();
        pingObservers();
    }

    public void run(Context context) {
        Log.v(TAG, getShortName() + "run");

        reset = false;
        running = true;
        initialized = true;

        makeUpdateTimestamp();
        pingObservers();
    }

    public void pause(Context context) {
        Log.v(TAG, getShortName() + "pause");

        running = false;
        initialized = true;

        makeUpdateTimestamp();
        pingObservers();
    }

    public void click(Context context) {
        Log.v(TAG, getShortName() + "click");
        if (isRunning())
            pause(context);
        else
            run(context);
    }

    public void pingObservers() {
        // this incantation will make observers elsewhere aware that there's new content
        Log.v(TAG, getShortName() + "pinging");
        setChanged();
        notifyObservers();
        clearChanged();
        Log.v(TAG, getShortName() + "ping complete");
    }

    /**
     * Return the time of either when the stopwatch began or when the countdown ends.
     * IF RUNNING, this time will be consistent with System.currentTimeMillis(), i.e., in GMT.
     * IF PAUSED, this time will be relative to zero and will be what should be displayed.
     * Make sure to call isRunning() to know how to interpret this result.
     * @return GMT time in milliseconds
     */
    abstract public long eventTime();

    /**
     * This converts an absolute time, as returned by eventTime, to a relative time
     * that might be displayed
     */
    public String relativeTimeString(long eventTime) {
        if(running) {
            long timeNow = System.currentTimeMillis();
            long delta = timeNow - eventTime;
            if (delta < 0) delta = -delta;
            return DateUtils.formatElapsedTime(delta / 1000);
        } else {
            if (eventTime < 0) eventTime = -eventTime;
            return DateUtils.formatElapsedTime(eventTime / 1000);
        }
    }

    public String toString() {
        return relativeTimeString(eventTime());
    }

    abstract public String getActionNotificationClickString();

    abstract public int getNotificationID();

    abstract public Class getActivity();

    abstract public int getIconID();

    abstract public String getShortName();
}

File: wear/src/main/kotlin/org/dwallach/xstopwatch/SharedState.kt
/*
 * XStopwatch / XTimer
 * Copyright (C) 2014 by Dan Wallach
 * Home page: http://www.cs.rice.edu/~dwallach/xstopwatch/
 * Licensing: http://www.cs.rice.edu/~dwallach/xstopwatch/licensing.html
 */
package org.dwallach.xstopwatch

import android.app.Activity
import android.content.Context
import android.text.format.DateUtils
import android.util.Log

import java.util.Observable

/**
 * We'll implement this abstract class for StopwatchState and TimerState.
 */
abstract class SharedState: Observable() {
    var isRunning: Boolean = false
        protected set
    var isReset: Boolean = true
        protected set
    var updateTimestamp: Long = 0 // when the last user interaction was
        protected set
    var isVisible: Boolean = false
        set(visible) {
            Log.v(TAG, "${shortName} visible: $visible")
            field = visible
            isInitialized = true

            makeUpdateTimestamp()
            pingObservers()
        }
    var isInitialized: Boolean = false
        protected set

    private fun makeUpdateTimestamp() {
        updateTimestamp = currentTime()
    }

    open fun reset(context: Context?) {
        Log.v(TAG, "${shortName} reset")
        isRunning = false
        isReset = true
        isInitialized = true

        makeUpdateTimestamp()
        pingObservers()
    }

    open fun run(context: Context) {
        Log.v(TAG, "${shortName} run")

        isReset = false
        isRunning = true
        isInitialized = true

        makeUpdateTimestamp()
        pingObservers()
    }

    open fun pause(context: Context) {
        Log.v(TAG, "${shortName} pause")

        isRunning = false
        isInitialized = true

        makeUpdateTimestamp()
        pingObservers()
    }

    fun click(context: Context) {
        Log.v(TAG, "${shortName} click")
        if (isRunning)
            pause(context)
        else
            run(context)
    }

    fun pingObservers() {
        // this incantation will make observers elsewhere aware that there's new content
        Log.v(TAG, "${shortName} pinging")
        setChanged()
        notifyObservers()
        clearChanged()
        Log.v(TAG, "${shortName} ping complete")
    }

    /**
     * Return the time of either when the stopwatch began or when the countdown ends.
     * IF RUNNING, this time will be consistent with System.currentTimeMillis(), i.e., in GMT.
     * IF PAUSED, this time will be relative to zero and will be what should be displayed.
     * Make sure to call isRunning() to know how to interpret this result.
     * @return GMT time in milliseconds
     */
    abstract fun eventTime(): Long

    /**
     * This converts an absolute time, as returned by eventTime, to a relative time
     * that might be displayed
     */
    fun relativeTimeString(eventTime: Long): String =
    DateUtils.formatElapsedTime(
        if (isRunning)
            Math.abs(currentTime() - eventTime) / 1000
        else
            Math.abs(eventTime) / 1000)

    override fun toString() = relativeTimeString(eventTime())

    abstract val actionNotificationClickString: String

    abstract val notificationID: Int

    abstract val activity: Class<out Activity>

    abstract val iconID: Int

    abstract val shortName: String

    companion object {
        private val TAG = "SharedState"

        fun currentTime() = System.currentTimeMillis()
    }
}
